### PR TITLE
Lambdamart implementation

### PR DESCRIPTION
**Pull request by srinikrish22**
I have implemented the LambdaMART algorithm for object ranking.  

## Description

I have created a class with the required core functionality of lambdamart and added ndcg metric implementation to the utils. I have added the ranker into the test_ranking and also added the ranker definition in the constants. 
## Motivation and Context

This change aims at adding an implementation of a new ranker for object ranking to the repository.  

## How Has This Been Tested?

I have added the possibility to test this ranker under the existing testing infrastructure of the cs-ranking repository. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: csrank/metrics.py [hunk 1]
```diff
@@ -123,3 +121,15 @@ def topk_acc(y_true, y_pred):
         return acc
 
     return topk_acc
+
+def dcg_at_k(r, k): #required for LambdaMART
```
Comment (kiudee) [inline csrank/metrics.py]: The comment can be removed here, if this is supposed to be a general purpose function.
Better would be a docstring, documenting what the function is computing.

File: csrank/metrics.py [hunk 2]
```diff
@@ -123,3 +121,15 @@ def topk_acc(y_true, y_pred):
...
+        r = np.asfarray(r)[:k]
+        if r.size:
+            return np.sum(np.subtract(np.power(2, r), 1) / np.log2(np.arange(2, r.size + 2)))
+        return 0.
+
+def ndcg_at_k(r, k):#required for LambdaMART
```
Comment (kiudee) [inline csrank/metrics.py]: see above

File: csrank/objectranking/lambdamart.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+from csrank.learner import Learner
+from csrank.metrics import ndcg_at_k
+from csrank.objectranking.object_ranker import ObjectRanker
+
+
+class LambdaMART(ObjectRanker,Learner):
+    def __init__(self, n_objects=None, n_object_features=None, number_of_trees=5, learning_rate=0.1,
+                 min_samples_split=2, max_depth=50, min_samples_leaf=1, max_leaf_nodes=None):
+        """
+            Create a LambdaMART based rank regression model. This model uses an ensemble of trees that learn to predict
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: It would be useful to add a reference to the paper in the docstring.
See [here](https://github.com/kiudee/cs-ranking/blob/4b11e43171c609a7219969a3a305128ac276d766/csrank/objectranking/expected_rank_regression.py#L58) for how to do that.

File: csrank/objectranking/lambdamart.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        self.number_of_trees = number_of_trees
+        self.learning_rate = learning_rate
+        self.min_samples_split = min_samples_split
+        self.max_depth = max_depth
+        self.min_samples_leaf = min_samples_leaf
+        self.max_leaf_nodes = max_leaf_nodes
+        self.trees = []
+        self.logger = logging.getLogger(LambdaMART.__name__)
+
+    def _custom_letor_dataset_reader(self, filepath):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: This should not be part of the learner.
The learner does not know, what datasets it is given.

File: csrank/objectranking/lambdamart.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        features = np.asarray(features_as_list)
+        scores_unflat = np.array(scores_as_list)
+        scores = np.reshape(scores_unflat,(len(scores_unflat),1))
+
+        #Concatenate the reshaped arrays and return as trainin data
+        train_data = np.concatenate((scores,features),axis=1)
+
+        return train_data
+
+    def _i_dcg(self,scores,k=None):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Why do you implement `_i_dcg` here and in `metrics.py`?

File: csrank/objectranking/lambdamart.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+        """
+            Returns the ideal DCG value for the given scores
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Missing input parameters and output description

File: csrank/objectranking/lambdamart.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+        """
+        scores = [score for score in sorted(scores)[::-1]]
+        if k == None:
+            return np.sum([(np.power(2, scores[i]) - 1) / np.log2(i + 2) for i in range(len(scores))])
+
+        elif k > 0:
+            return np.sum([(np.power(2, scores[i]) - 1) / np.log2(i + 2) for i in range(len(scores[:k]))])
+
+    def _group_queries(self, training_data, qid_index):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Docstring could be useful

File: csrank/objectranking/lambdamart.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            Parameters
+            ----------
+            X : numpy array
+                (n_instances, n_objects, n_features)
+                Feature vectors of the objects
+            Y : numpy array
+                (n_instances, n_objects)
+                Rankings of the given objects
+            **kwargs
+                Keyword arguments for the fit function
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: `**kwargs` is not used in the function. Maybe pass it on to the `DecisionTreeRegressor` and document it here?

File: csrank/objectranking/lambdamart.py [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+                                         min_samples_split=self.min_samples_split,
+                                         min_samples_leaf=self.min_samples_leaf,
+                                         min_weight_fraction_leaf=0.,
+                                         max_features=None,
+                                         random_state=9,
+                                         max_leaf_nodes=self.max_leaf_nodes,
+                                         min_impurity_decrease=0.,
+                                         min_impurity_split=None)
+            tree.fit(training_data[:,2:], lambdas)
+            self.trees.append(tree)
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: `self.trees` is not cleared - thus every time you call `fit`, it will keep appending new trees here.
This is also a nice candidate for a test.

File: csrank/objectranking/lambdamart.py [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+    def predict_for_scores(self, scores, **kwargs):
+        return ObjectRanker.predict_for_scores(self, scores, **kwargs)
+
+    def predict(self, X, **kwargs):
+        return super().predict(X, **kwargs)
+    
+    def _predict(self, data):
+        """
+            Predict the
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: "Predict the ..." ?

File: csrank/objectranking/lambdamart.py [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+                                         random_state=9,
+                                         max_leaf_nodes=self.max_leaf_nodes,
+                                         min_impurity_decrease=0.,
+                                         min_impurity_split=None)
+            tree.fit(training_data[:,2:], lambdas)
+            self.trees.append(tree)
+            prediction = tree.predict(training_data[:,2:])
+            predicted_scores += prediction * self.learning_rate
+    
+    def _predict_scores_fixed(self, X, **kwargs):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Docstring?

File: csrank/objectranking/lambdamart.py [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+        n_instances, n_objects, n_features = X.shape
+        self.logger.info("For Test instances {} objects {} features {}".format(*X.shape))
+        X1 = X.reshape(n_instances * n_objects, n_features)
+        scores = np.zeros(n_instances * n_objects)
+        for tree in self.trees:
+            scores += tree.predict(X1)
+        scores = scores.reshape(n_instances, n_objects)
+        return scores
+
+    def predict_scores(self, X, **kwargs):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Docstring

File: csrank/objectranking/lambdamart.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+        return super().predict_scores(X, **kwargs)
+
+    def predict_for_scores(self, scores, **kwargs):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Docstring

File: csrank/objectranking/lambdamart.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+        return ObjectRanker.predict_for_scores(self, scores, **kwargs)
+
+    def predict(self, X, **kwargs):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Docstring

File: csrank/objectranking/lambdamart.py [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+        query_indexes = self._group_queries(data, 0)
+        predicted_scores = np.zeros(len(data))
+        for query in query_indexes:
+            results = np.zeros(len(query_indexes[query]))
+            for tree in self.trees:
+                results += self.learning_rate * tree.predict(data[query_indexes[query], 1:])
+            predicted_scores[query_indexes[query]] = results
+        return predicted_scores
+
+    def _validate(self, data, k=10):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Docstring

The model should only be validated from the outside. This method can be removed.

File: csrank/objectranking/lambdamart.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+        """
+        self.min_samples_split = min_samples_split
+        self.max_depth = max_depth
+        self.min_samples_leaf = min_samples_leaf
+        self.max_leaf_nodes = max_leaf_nodes
+        self.number_of_trees = number_of_trees
+        self.learning_rate = learning_rate
+
+
+def compute_lambda_weights(args):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Should this be called by the user or can this be `_compute_lambda_weights`.

File: csrank/objectranking/lambdamart.py [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+    """
+        This is used by the LambdaMART learner. Computes the lambda targets that are considered to be the relationship between the different features and the ranking of a document. These lambdas are the learning target for the base learner.
+        The parameters to this function are passed in the form of a zipped iterator.
+
+        Parameters
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Those are not the parameters of the function. It only receives args (which is not good).

File: csrank/objectranking/lambdamart.py [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+        lambdas[j] -= lambda_val
+
+        w_val = rho * rho_complement * z_ndcg
+        w[i] += w_val
+        w[j] += w_val
+
+    return lambdas[rev_indexes], w[rev_indexes], query_key
+
+
+def _single_dcg(scores, i, j):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: See remark above

File: csrank/objectranking/lambdamart.py [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+from sklearn.tree import DecisionTreeRegressor
+
+from csrank.learner import Learner
+from csrank.metrics import ndcg_at_k
+from csrank.objectranking.object_ranker import ObjectRanker
+
+
+class LambdaMART(ObjectRanker,Learner):
+    def __init__(self, n_objects=None, n_object_features=None, number_of_trees=5, learning_rate=0.1,
+                 min_samples_split=2, max_depth=50, min_samples_leaf=1, max_leaf_nodes=None):
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Add a **kwargs here.

File: csrank/objectranking/lambdamart.py [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+
+            #Hyperparameters to be considered are :"min_samples_split","max_depth","min_samples_leaf","max_leaf_nodes"
+            tree = DecisionTreeRegressor(criterion="mse",
+                                         splitter="best",
+                                         max_depth=self.max_depth,
+                                         min_samples_split=self.min_samples_split,
+                                         min_samples_leaf=self.min_samples_leaf,
+                                         min_weight_fraction_leaf=0.,
+                                         max_features=None,
+                                         random_state=9,
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: The random_state should be passed in `__init__`.
See here: https://github.com/kiudee/cs-ranking/blob/4b11e43171c609a7219969a3a305128ac276d766/csrank/objectranking/expected_rank_regression.py#L67

File: csrank/objectranking/lambdamart.py [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+            # filename = str("lmartDebug2_"+str(k)+"_lambdas.txt")
+            # np.savetxt(filename, lambdas)
+
+            #Hyperparameters to be considered are :"min_samples_split","max_depth","min_samples_leaf","max_leaf_nodes"
+            tree = DecisionTreeRegressor(criterion="mse",
+                                         splitter="best",
+                                         max_depth=self.max_depth,
+                                         min_samples_split=self.min_samples_split,
+                                         min_samples_leaf=self.min_samples_leaf,
+                                         min_weight_fraction_leaf=0.,
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: All parameters here should be passable in `__init__` and tunable.

File: csrank/objectranking/lambdamart.py [hunk 22]
```diff
@@ -0,0 +1,370 @@
...
+            w = np.zeros(len(predicted_scores))
+            pred_scores = [predicted_scores[query_indexes[query]] for query in query_keys]
+            
+            pool = Pool()
+            for lambda_val, w_val, query_key in pool.map(compute_lambda_weights, zip(true_scores, pred_scores, sorted_query_pair, idcg, query_keys), chunksize=1):
+                indexes = query_indexes[query_key]
+                lambdas[indexes] = lambda_val
+                w[indexes] = w_val
+            pool.close()
+            # filename = str("lmartDebug2_"+str(k)+"_lambdas.txt")
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: Remove this comment

File: csrank/objectranking/lambdamart.py [hunk 23]
```diff
@@ -0,0 +1,370 @@
...
+
+        idcg = [self._i_dcg(scores) for scores in true_scores]
+
+        for k in range(self.number_of_trees):
+            print("Tree %d" % (k))
+            lambdas = np.zeros(len(predicted_scores))
+            w = np.zeros(len(predicted_scores))
+            pred_scores = [predicted_scores[query_indexes[query]] for query in query_keys]
+            
+            pool = Pool()
```
Comment (kiudee) [inline csrank/objectranking/lambdamart.py]: If you use a pool for multiprocessing, make sure that the user can pass the number of processes it `fit`.
Also, use a `with` environment like so:
```python
with Pool(5) as p:
    print(p.map(f, [1, 2, 3]))
```

Maybe even use joblib.

**Comment by kiudee** [comment]
Thank you for the new work on the implementation. Could you rebase your branch onto master and force push it here? Then it would be easier to review the changes.

**Comment by prithagupta** [comment]
@srinikrish22 I would agree with @kiudee. 
Secondly, regarding the tests, I would suggest that you follow the pull request https://github.com/kiudee/cs-ranking/pull/80
run the test locally to just check the environment problem. 
The current error is due coverage, which should not have any effect on your implementation work

**Comment by srinikrish22** [comment]
I have finished the rebase @kiudee. I had done it with my master but forgot to rebase my branch on my master. Now it is be done. 
@prithagupta I tried to run the tests locally but my tensorflow installation is broken. I am currently trying to get the tests running on a colab instance.

**Comment by prithagupta** [comment]
@srinikrish22 Dd you update the tensorflow, the 2.0 version should have a lot of problems. as far as i know, your ranker does not use tensorflow. So you should just fix your version of the last stable release of the tensorflow.

**Comment by kiudee** [comment]
@srinikrish22 I think you need to first pull the latest master before you do the rebase. Currently, there are still commits in this branch which do not belong to your changes (example https://github.com/kiudee/cs-ranking/pull/52/commits/6c4c30cec89363b97d5c2c238cd0a77637b7a4ee)

**Comment by srinikrish22** [comment]
@kiudee I did rebase on the latest master branch recently. Those changes are because of the merge conflicts from the rebase.

**Comment by srinikrish22** [comment]
@kiudee I have created a new pull request to clarify the confusion about rebase with the current master. This should make things easier to compare the changes.

**Comment by timokau** [comment]
Closing this in favor of https://github.com/kiudee/cs-ranking/pull/82.
